csrc/expr_evaluator.cpp

```diff
@@ -344,6 +344,143 @@ void ExpressionEvaluator::print() const {
+    for (auto [dim_size_2, tv_id_pairs_2] : size_to_info) {
+      if (dim_size_1 <= dim_size_2) {
+        // N^2 algorithm, only process when one size is less than the other,
+        // avoids duplicate entries.
+        continue;
+      }
+
+      bool tv1_is_consumer = false;
+      for (auto tv_id_pair_1 : tv_id_pairs_1) {
+        for (auto tv_id_pair_2 : tv_id_pairs_2) {
+          // Check for producer-consumer relationship
+          auto producer_tvs_of_tv1 =
+              ir_utils::producerTvsOf(tv_id_pair_1.first);
+          auto producer_tvs_of_tv2 =
+              ir_utils::producerTvsOf(tv_id_pair_2.first);
+          if (std::find(
+                  producer_tvs_of_tv1.begin(),
+                  producer_tvs_of_tv1.end(),
+                  tv_id_pair_2.first) != producer_tvs_of_tv1.end()) {
+            tv1_is_consumer = true;
+          } else if (
+              std::find(
+                  producer_tvs_of_tv2.begin(),
+                  producer_tvs_of_tv2.end(),
+                  tv_id_pair_1.first) == producer_tvs_of_tv2.end()) {
+            // No relationship found, skip
+            continue;
+          }
+
+          Expr* relationship = tv1_is_consumer
+              ? tv_id_pair_1.first->definition()
+              : tv_id_pair_2.first->definition();
+
+          // Found at least one consumer/producer relationship with mismatched
+          // sizes.
+          found_producer_consumer_issue = true;
+
+          // Produce error message. Normally I'd just use swap but keys in an
+          // unordered map are const, so doing some juggling with the strings
+          // instead.
+          std::stringstream tv1_error;
+          std::stringstream tv2_error;
+          tv1_error << " TV: " << tv_id_pair_1.first
+                    << " id: " << tv_id_pair_1.second
+                    << " found size: " << dim_size_1 << "\n";
+          tv2_error << " TV: " << tv_id_pair_2.first
+                    << " id: " << tv_id_pair_2.second
+                    << " found size: " << dim_size_2 << "\n";
+          err_msg << "  For Producer"
+                  << (tv1_is_consumer ? tv2_error.str() : tv1_error.str());
+          err_msg << "  For Consumer"
+                  << (tv1_is_consumer ? tv1_error.str() : tv2_error.str());
+          err_msg
+              << "  With producer-consumer relationship through the expression: "
+              << relationship << "\n";
+        }
+      }
+    }
+  }
+
+  if (found_producer_consumer_issue) {
+    NVF_THROW(err_msg.str());
+  }
+
+  if (size_to_info.size() > 1) {
+    for (const auto& [size, info_pairs] : size_to_info) {
+      err_msg << "Size " << size << " found for ID, in TV:\n";
+      for (auto info_pair : info_pairs) {
+        err_msg << "  " << info_pair.second << ", " << info_pair.first << "\n";
+      }
+    }
+    err_msg << "These sizes should all match.\n";
+  } else {
+    std::unordered_map<int64_t, std::vector<const IterDomain*>> size_to_ids;
```

Comment:
This branch will never happen, correct? AFAIU, if `size_to_info.size()==1`, all IterDomains in `id_set` have the same extent and the check will pass.

Reply:
Yes, that should hold given the only way we could end up with IDs in the set that don't match here is if there's IDs registered that are not within the TensorViews of the Fusion. Technically possible, but that's an unlikely thing to happen.